src/blog/0.3/linear-types-borrowing.vmd

```diff
@@ -0,0 +1,673 @@
+
+
+[# E: How is this next part, where we blend gen refs back in? I tried to find an elegant way to build up to "the solution was with us all along".]
+
+The other solution, surprisingly, is to blend generational references back in.
+
+Each object is still owned linearly; only one place can own an object at a time. But we'll also allow a generational reference to point at the object.
+
+With this, we can address a lot of the drawbacks:
+
+ 1. We can have as many references to data as we want, though there is still only one owner.
+ 1. We don't need to restructure our program to accommodate them.
+ 1. We can use patterns like intrusive data structures and graphs. [# In Vale, we would still have all nodes owned by a central collection, but we could still have references between nodes just fine.]
+ 1. We can use observers, back-references, dependency references, callbacks, delegates, whatever we want.
+ 1. We still don't have data races, because Vale changes the generations of any objects that cross region boundaries (unless we safely avoid that with [isolates](/blog/zero-cost-borrowing-regions-part-2-isolates)).
+
+
+Generational references seem to occupy a sweet spot because they allow objects to be linear (they're owned by only one place), yet allow shared mutability in a way that doesn't artificially extend the lifetime of the object.
+
+
+Since we're combining linear types with aliasing, we'll call this the *linear-aliasing model*.
+
+
+The mark of a good language is when its features compose well and don't conflict with each other. It seems we're on the right track with this model, as it composes linear types with shared mutability pretty well.
+
+
+<slice/>
+
+# Matches Made in Heaven
+
+A pleasant surprise, we're back to where we started!
+
+
+But we arrive with a new understanding:
+
+ * We can code in a "linear-first" fashion, and then only add references if we want some extra flexibility.
+ * Alternatively, we can code with a lot of references, and then bring them down to *zero* where we want some more speed.
+
+In other words, we now have complete control over the flexibility vs. performance tradeoff, something I've never seen a language do well.
+
+
+It was quite a shock, realizing that Vale could do this the entire time.
+
+
+And it made me wonder, what does this mean for [regions](/blog/zero-cost-borrowing-regions-overview), Vale's other mechanism for eliminating generation checks?
+
+
+Then I realized, they fit _perfectly_ together. We can use regions to look at a function's inputs in a zero-cost way, and use this "linear style" for any new data the function creates.
+
+I tried this out in a little [roguelike game](https://github.com/Verdagon/RegionsBenchmarks/blob/main/tiny-roguelike/roguelike.vale). The vast majority of generation checks were eliminated by regions, and the rest were eliminated by this new linear style.
+
+[# E: Should I go into more detail about this balance, or save it for a future article?]
```

Comment:
Depends how long it would be, I'd say you're at length for this article so save it, but if it's really short and not worth its own article then add it here